File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys

def isCheck(s):
    stack = [0]   # [-1] 할때 오류 안나기 위해서
    for i in s:
        if i in {'[', '('}: stack.append(i)

        elif i == ']':
            if stack[-1] == '[': stack.pop()
            else               : return 'no'
        
        elif i == ')':
            if stack[-1] == '(': stack.pop()
            else               : return 'no'
    
    return 'yes' if stack == [0] else 'no'

while(1):
    s = sys.stdin.readline().rstrip()
    if s == '.': break
    print(isCheck(s))

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <iostream>
#include <stack>
using namespace std;

int main(void) {
    while (1) {
        string s;
        getline(cin, s);
        if (s == ".") break;

        stack<char> a;
        bool flag = true;
        for (auto i : s) {
            if (i == '(' || i == '[') a.push(i);

            else if (i == ')') {
                if (a.empty() || a.top() != '(') {
                    flag = false;
                    break;
                }
                a.pop();
            }

            else if (i == ']') {
                if (a.empty() || a.top() != '[') {
                    flag = false;
                    break;
                }
                a.pop();
            }
        }
        if (flag && a.empty()) cout << "yes\n";
        else cout << "no\n";
    }
    return 0;
}